src/clib/pio_spmd.c

```diff
@@ -472,50 +407,50 @@ int pio_swapm(void *sndbuf, int *sndlths, int *sdispls, MPI_Datatype *stypes,
     for (istep = 0; istep < steps; istep++)
     {
         p = swapids[istep];
-        if (sndlths[p] > 0)
-	{
-            tag = mytask + offset_t;
+        if (sendcounts[p] > 0)
+        {
+            tag = my_rank + offset_t;
             if (handshake)
-	    {
-                CheckMPIReturn(MPI_Wait(hs_rcvids+istep, &status), __FILE__,__LINE__);
+            {
+                CheckMPIReturn(MPI_Wait(hs_rcvids + istep, &status), __FILE__, __LINE__);
```

Comment:
If handshake is enabled don't post sends until the receiving task has posted recvs